File: src/hwacc/AcceleratorIommu.py
# Copyright (c) 2026 SALAM contributors
# SPDX-License-Identifier: BSD-3-Clause
#
# AcceleratorIommu -- a single, device-wide IOMMU/SMMU SimObject.
#
# One instance is shared by every LLVMInterface (CU) and CommInterface
# inside an AccCluster: the accelerator device is modeled as ONE unit
# sitting behind ONE chip-wide SMMU. The CUs consult this object to
# translate addresses; the object owns the IOTLB cache and the
# in-order translation port queue.
#
# Replaces the file-scope statics that used to live in LLVMInterface
# and CommInterface (iotlbLru/iotlbSet, sIommuNextReadyTick).
from m5.params import *
from m5.proxy import *
from m5.SimObject import SimObject


class AcceleratorIommu(SimObject):
    type = "AcceleratorIommu"
    cxx_class = "gem5::AcceleratorIommu"
    cxx_header = "hwacc/accelerator_iommu.hh"

    # Master toggle. When False the CU short-circuits its translate()
    # call and behaves bit-identically to plain mode (no overhead, no
    # bookkeeping). The IOMMU object itself is always instantiated so
    # that LLVMInterface's `Parent.any` proxy can resolve, regardless
    # of whether IOMMU mode is active.
    enabled = Param.Bool(False, "Master IOMMU toggle")

    # IOTLB geometry. Defaults are tuned to a low-power IoT peripheral
    # IOMMU (Arm MMU-400 class):
    #   * 8-entry uTLB         -- silicon-area constrained vendor IPs
    #                             ship 4-16 entries
    #   * 2 ns hit latency     -- ~1 cycle SRAM lookup @ 400-500 MHz
    #   * 500 ns miss latency  -- 4-level walk to slow DRAM with no
    #                             walk caches and a single PTW thread
    iotlb_entries     = Param.UInt32(8,        "IOTLB capacity (LRU)")
    hit_latency       = Param.Tick(2000,       "IOTLB hit latency (ticks)")
    miss_latency      = Param.Tick(500000,     "IOTLB miss / page-walk "
                                               "latency (ticks)")

File: src/hwacc/accelerator_iommu.cc
/*
 * Copyright (c) 2026 SALAM contributors
 * SPDX-License-Identifier: BSD-3-Clause
 */
#include "hwacc/accelerator_iommu.hh"

#include <algorithm>

namespace gem5
{

AcceleratorIommu::AcceleratorIommu(const Params &p)
    : SimObject(p),
      enabledFlag(p.enabled),
      iotlbEntries(p.iotlb_entries),
      hitLatency(p.hit_latency),
      missLatency(p.miss_latency),
      nextReadyTick(0),
      stats(this)
{
}

bool
AcceleratorIommu::iotlbAccess(uint64_t pageAddr)
{
    if (iotlbEntries == 0) {
        // Caching disabled: still record the access in stats but
        // always miss.
        return false;
    }

    auto it = set.find(pageAddr);
    if (it != set.end()) {
        // Hit. Promote to MRU.
        lru.remove(pageAddr);
        lru.push_front(pageAddr);
        return true;
    }

    // Miss. Insert; evict LRU if at capacity.
    lru.push_front(pageAddr);
    set.insert(pageAddr);
    if (lru.size() > iotlbEntries) {
        uint64_t evicted = lru.back();
        lru.pop_back();
        set.erase(evicted);
    }
    return false;
}

Tick
AcceleratorIommu::translate(uint64_t pageAddr, bool /*isRead*/)
{
    // Lat=0 / disabled fast path: zero overhead, identical to plain.
    if (!enabledFlag || (hitLatency == 0 && missLatency == 0)) {
        return curTick();
    }

    bool hit = iotlbAccess(pageAddr);
    Tick lat = hit ? hitLatency : missLatency;

    // Single chip-wide translation port: serialize. Each access
    // begins no earlier than max(curTick(), nextReadyTick), then
    // occupies the port for `lat`. Subsequent accesses queue.
    Tick startTick = std::max(curTick(), nextReadyTick);
    Tick ready     = startTick + lat;
    nextReadyTick  = ready;

    stats.totalChecks++;
    if (hit)
        stats.tlbHits++;
    else
        stats.tlbMisses++;
    stats.totalLatencyTicks += lat;

    return ready;
}

AcceleratorIommu::IommuStats::IommuStats(statistics::Group *parent)
    : statistics::Group(parent),
      ADD_STAT(totalChecks,       statistics::units::Count::get(),
               "IOTLB lookups performed"),
      ADD_STAT(tlbHits,           statistics::units::Count::get(),
               "IOTLB hits"),
      ADD_STAT(tlbMisses,         statistics::units::Count::get(),
               "IOTLB misses (page walks)"),
      ADD_STAT(totalLatencyTicks, statistics::units::Tick::get(),
               "Sum of latency added by IOMMU (ticks)")
{
}

} // namespace gem5
